src/main/java/com/edmunds/rest/databricks/service/GroupsServiceImpl.java

```diff
@@ -0,0 +1,149 @@
+package com.edmunds.rest.databricks.service;
+
+import com.edmunds.rest.databricks.DTO.PrincipalNameDTO;
+import com.edmunds.rest.databricks.DatabricksRestException;
+import com.edmunds.rest.databricks.RequestMethod;
+import com.edmunds.rest.databricks.restclient.DatabricksRestClient;
+import com.fasterxml.jackson.core.type.TypeReference;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.log4j.Logger;
+
+/**
+ * The implementation of GroupsService.
+ */
+public class GroupsServiceImpl extends DatabricksService implements GroupsService {
+
+  private static Logger log = Logger.getLogger(GroupsServiceImpl.class);
+  private static final PrincipalNameDTO[] EMPTY_MEMBERS_ARRAY = {};
+  private static final String[] EMPTY_STRING_ARRAY = {};
+
+  public GroupsServiceImpl(final DatabricksRestClient client) {
+    super(client);
+  }
+
+  @Override
+  public void addUserToGroup(String userName, String parentName) throws IOException, DatabricksRestException {
+    Map<String, Object> data = new HashMap<>();
+    data.put("user_name", userName);
+    data.put("parent_name", parentName);
+    client.performQuery(RequestMethod.POST, "/groups/add-member", data);
+  }
+
+  @Override
+  public void addGroupToGroup(String groupName, String parentName) throws IOException, DatabricksRestException {
+    Map<String, Object> data = new HashMap<>();
+    data.put("group_name", groupName);
+    data.put("parent_name", parentName);
+    client.performQuery(RequestMethod.POST, "/groups/add-member", data);
+  }
+
+  @Override
+  public String createGroup(String groupName) throws IOException, DatabricksRestException {
+    boolean groupExists = groupExists(groupName);
+    // Doing this check and logging rather than letting API throw a "RESOURCE_ALREADY_EXISTS" exception
+    if (!groupExists) {
+      Map<String, Object> data = new HashMap<>();
+      data.put("group_name", groupName);
+      byte[] responseBody = client.performQuery(RequestMethod.POST, "/groups/create", data);
+      Map<String, Object> response = this.mapper.readValue(responseBody, Map.class);
+      return (String) response.get("group_name");
```

Comment:
is there any need to do null checking here?

Reply:
hmm, so I think if it reaches this block of code, that means there is no group with that name yet, so databricks will create it and return the name back to you, i.e. in the typical use case this method returns the same string that was passed in. However, if something wonky happens and databricks doesn't return the group name that was created (causing an NPE), I think we should let it fail because something obviously went wrong and I don't know how to recover from it.

Reply:
yeah I think the latter is a good idea